c10/cuda/impl/CUDAGuardImpl.h

```diff
@@ -29,20 +29,17 @@ struct CUDAGuardImpl final : public c10::impl::DeviceGuardImplInterface {
   }
   Device exchangeDevice(Device d) const override {
     TORCH_INTERNAL_ASSERT(d.is_cuda());
-    Device old_device = getDevice();
-    if (old_device.index() != d.index()) {
-      C10_CUDA_CHECK(cudaSetDevice(d.index()));
-    }
-    return old_device;
+    int old_device_index = c10::cuda::ExchangeDevice(d.index());
+    return Device(DeviceType::CUDA, old_device_index);
```

Comment:
this creates new `Device` object even in the common case of requested device being the same as current device (which wasn't the case previously), can you avoid this? If not, that's fine, that's relatively minor.